File: app/src/main/java/igrek/forceawaken/service/volume/NoiseDetectorService.java
package igrek.forceawaken.service.volume;

import android.media.MediaRecorder;
import android.os.Handler;

import java.io.IOException;

import igrek.forceawaken.logger.Logger;

public class NoiseDetectorService {
	
	private Logger logger;
	private MediaRecorder mediaRecorder;
	
	public NoiseDetectorService(Logger logger) {
		this.logger = logger;
	}
	
	public void measureNoiseLevel(long millis, NoiseLevelMeasureCallback completeCallback) {
		// measure surrounding loudness level
		try {
			mediaRecorder = new MediaRecorder();
			mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
			mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
			mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
			mediaRecorder.setOutputFile("/dev/null");
			mediaRecorder.prepare();
			mediaRecorder.start();
			mediaRecorder.getMaxAmplitude(); // initialize measurement
			
			new Handler().postDelayed(() -> {
				
				try {
					int amplitude = mediaRecorder.getMaxAmplitude();
					double amplitudeDb = 20 * Math.log10((double) Math.abs(amplitude));
					logger.info("Surrounding noise amplitude: " + amplitudeDb + " dB");
					mediaRecorder.stop();
					mediaRecorder.release();
					completeCallback.onComplete(amplitudeDb);
				} catch (Exception e) {
					logger.error(e);
					completeCallback.onComplete(0);
				}
				
			}, millis);
			
		} catch (IOException | IllegalStateException e) {
			logger.error(e);
			completeCallback.onComplete(0);
		}
	}
	
	public interface NoiseLevelMeasureCallback {
		void onComplete(double amplitudeDb);
	}
	
}

File: app/src/main/java/igrek/forceawaken/volume/NoiseDetectorService.kt
package igrek.forceawaken.volume

import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import igrek.forceawaken.info.logger.Logger
import igrek.forceawaken.info.logger.LoggerFactory
import java.io.IOException
import kotlin.math.abs
import kotlin.math.log10

class NoiseDetectorService() {

    private val logger: Logger = LoggerFactory.logger

    fun measureNoiseLevel(millis: Long, completeCallback: NoiseLevelMeasureCallback) {
        // measure surrounding loudness level
        try {
            val mediaRecorder = MediaRecorder()
            mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC)
            mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP)
            mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB)
            mediaRecorder.setOutputFile("/dev/null")
            mediaRecorder.prepare()
            mediaRecorder.start()
            mediaRecorder.maxAmplitude // initialize measurement
            Handler(Looper.getMainLooper()).postDelayed({
                try {
                    val amplitude: Int = mediaRecorder.maxAmplitude
                    val amplitudeDb = 20 * log10(abs(amplitude).toDouble())
                    logger.info("Surrounding noise amplitude: $amplitudeDb dB")
                    mediaRecorder.stop()
                    mediaRecorder.release()
                    completeCallback.onComplete(amplitudeDb)
                } catch (e: Exception) {
                    logger.error(e)
                    completeCallback.onComplete(0.0)
                }
            }, millis)
        } catch (e: IOException) {
            logger.error(e)
            completeCallback.onComplete(0.0)
        } catch (e: IllegalStateException) {
            logger.error(e)
            completeCallback.onComplete(0.0)
        }
    }

    interface NoiseLevelMeasureCallback {
        fun onComplete(amplitudeDb: Double)
    }
}